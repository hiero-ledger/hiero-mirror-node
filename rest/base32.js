// SPDX-License-Identifier: Apache-2.0

import {base32} from 'rfc4648';

const decodeOpts = {loose: true};
const encodeOpts = {pad: false};

/**
 * Decodes the rfc4648 base32 string into a {@link Uint8Array}. If the input string is null, returns null.
 * @param str the base32 string.
 * @return {Uint8Array}
 */
const decode = (str) => str && base32.parse(str, decodeOpts);

/**
 * Encodes the byte array into a rfc4648 base32 string without padding. If the input is null, returns null. Note with
 * the rfc4648 loose = true option, it allows lower case letters, padding, and auto corrects 0 -> O, 1 -> L, 8 -> B
 * @param {Buffer|Uint8Array} data
 * @return {string}
 */
const encode = (data) => data && base32.stringify(data, encodeOpts);

export default {
  decode,
  encode,
};
