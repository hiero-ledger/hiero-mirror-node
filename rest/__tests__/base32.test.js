// SPDX-License-Identifier: Apache-2.0

import base32 from '../base32';

describe('decode', () => {
  describe('valid', () => {
    test('32W35LY', () => {
      expect(base32.decode('32W353Y')).toEqual(Uint8Array.from(Buffer.from('deadbeef', 'hex')));
    });
    test('32W35366', () => {
      expect(base32.decode('32W35366')).toEqual(Uint8Array.from(Buffer.from('deadbeefde', 'hex')));
    });
    test('null', () => {
      expect(base32.decode(null)).toBeNull();
    });
  });

  describe('invalid', () => {
    const invalidBase32Strs = [
      // A base32 group without padding can have 2, 4, 5, 7 or 8 characters from its alphabet
      'A',
      'AAA',
      'AAAAAA',
      // non-base32 characters, note due to the loose option, 0, 1, and 8 will be auto corrected to O, L, and B
      '9',
    ];
    invalidBase32Strs.forEach((invalidBase32Str) => {
      test(`${invalidBase32Str}`, () => {
        expect(() => base32.decode(invalidBase32Str)).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

describe('encode', () => {
  test('0xdeadbeef', () => {
    expect(base32.encode(Buffer.from('deadbeef', 'hex'))).toBe('32W353Y');
  });
  test('0xdeadbeefde', () => {
    expect(base32.encode(Buffer.from('deadbeefde', 'hex'))).toBe('32W35366');
  });
  test('null', () => {
    expect(base32.encode(null)).toBeNull();
  });
});
