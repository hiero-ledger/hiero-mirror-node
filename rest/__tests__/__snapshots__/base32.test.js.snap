// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`decode invalid 9 1`] = `"Invalid character 9"`;

exports[`decode invalid A 1`] = `"Unexpected end of data"`;

exports[`decode invalid AAA 1`] = `"Unexpected end of data"`;

exports[`decode invalid AAAAAA 1`] = `"Unexpected end of data"`;
